mtags/src/main/scala/scala/meta/internal/pc/TypeDefinitionProvider.scala

```diff
@@ -0,0 +1,136 @@
+package scala.meta.internal.pc
+
+import org.eclipse.{lsp4j => l}
+import scala.meta.internal.mtags.MtagsEnrichments._
+import scala.meta.pc.OffsetParams
+import scala.meta.internal.jdk.CollectionConverters._
+import scala.meta.tokenizers.Api
+
+class TypeDefinitionProvider(val compiler: MetalsGlobal) extends Api {
+  import compiler._
+
+  val ignoredTags: List[String] = List(
+    "val",
+    "var"
+  )
+
+  def tree(params: OffsetParams): Option[Tree] =
+    if (params.isWhitespace)
+      None
+    else {
+      val (_, _, tree) = createCompilationUnit(params)
+      Some(tree)
+    }
+
+  def lineAndCharacterToOffset(
+      pos: compiler.Position,
+      line: Int,
+      char: Int
+  ): Int =
+    pos.source.lineToOffset(line) + char
+
+  def typeSymbol(params: OffsetParams): Option[Symbol] = {
+    tree(params)
+      .flatMap {
+        case vd: ValDef =>
+          val (rStart, rEnd) = (vd.pos.start, vd.pos.end)
+          val tokens = vd.pos.source.content
+            .slice(rStart, rEnd)
+            .mkString
+            .tokenize
+            .get
+            .tokens
+          val curToken = tokens
+            .dropWhile(_.pos.end < params.offset() - vd.pos.start)
+            .head
+
+          if (ignoredTags.contains(curToken.text))
+            None
+          else
+            Some(vd)
+        case t: Tree => Some(t)
+      }
+      .flatMap {
+        case sel: Select
+            if sel.symbol.asMethod.returnType.typeSymbol.isTypeParameter =>
+          Some(sel.tpe.typeSymbol)
+        case app @ Apply(fun, args) if args.nonEmpty =>
+          val (rStart, rEnd) = (fun.pos.start, args.last.pos.end)
+
+          val txt = app.pos.source.content
+            .slice(rStart, rEnd)
+            .mkString
+          txt.tokenize.get.tokens.toList
+            .find(t => {
+              val (pStart, pEnd) =
+                (app.pos.start + t.pos.start, app.pos.start + t.pos.end)
+              pStart <= params.offset && pEnd >= params.offset
+            }) match {
+            case Some(t) =>
+              app.symbol.asMethod.paramLists.flatten
+                .find(_.nameString.trim == t.text)
+                .map(_.tpe.typeSymbol)
+            case _ => None
```

Comment:
maybe fixing `typedHoverTreeAt` is possible to untangle this code a bit?
Also, @sswistun-vl  [mentioned](https://github.com/scalameta/metals/pull/978#issuecomment-562541418) solution for problem of named parameter in [semanticDB](https://github.com/scalameta/scalameta/pull/1919). Not sure though, how it can save working with compiler here.